src/classes/CW_CampaignTriggerHandler.cls

```diff
@@ -0,0 +1,308 @@
+    	for (CW_Schedule__c s :schedules) {
+    		List<CW_Schedule__c> campaignSchedules = campaignScheduleMap.get(s.Marketing_Campaign__c);
+    		if (campaignSchedules==null) campaignSchedules = new List<CW_Schedule__c>();
+    		campaignSchedules.add(s);
+    		campaignScheduleMap.put(s.Marketing_Campaign__c,campaignSchedules);
+    	}
+    	
+    	// update schedules with Campaign stage value
+    	List<CW_Schedule__c> schedulesToUpdate = new List<CW_Schedule__c>();
+    	for (CW_Campaign__c c :campaigns.values()) {
+    		List<CW_Schedule__c> campaignSchedules = campaignScheduleMap.get(c.Id);
+    		
+    		if (campaignSchedules!=null) {
+	    		for (CW_Schedule__c s :campaignSchedules) {
+	    			s.Campaign_Stage__c = c.Stage__c;
+	    			schedulesToUpdate.add(s);
+	    		}
+    		}
+    	}
+    	
+    	if (!schedulesToUpdate.isEmpty()) update schedulesToUpdate;
+    }
+	
+	public void stopNewCampaignCreationWhenRoadmapLocked(List<CW_Campaign__c> newObjects)
+	{
+		List<CW_Roadmap__c> roadmapDetails = [SELECT Id, Approval_Status__c
+											  FROM CW_Roadmap__c
+											  WHERE Id =: newObjects[0].Roadmap__c];
+											  
+		if (!roadmapDetails.isEmpty())
+		{
+			for (CW_Roadmap__c curr : roadmapDetails)
+			{
+				if (curr.Approval_Status__c == ConstantUtil.APPROVAL_PROCESS_APPROVED ||
+					curr.Approval_Status__c == ConstantUtil.APPROVAL_PROCESS_IN_APPROVAL)
+				{
+					newObjects[0].addError(ConstantUtil.BLOCKED_NEW_CAMPAIGN);
+					break;
+				}
+			}
+		}
+	}
+	
+	public void createSingleSchedules(List<CW_Campaign__c> newCamps)
+	{
+		List<CW_Schedule__c> schs = new List<CW_Schedule__c>();
+		for(CW_Campaign__c camp : newCamps)
+		{
+			if(camp.Recurrence_Type__c == ConstantUtil.CAMPAIGN_RECURRENCE_TYPE_SINGLE)
+			{
+				CW_Schedule__c sch = new CW_Schedule__c();
+				sch.Marketing_Campaign__c = camp.id;
+				sch.Start_Date__c = camp.Campaign_Start_Date__c;
+				sch.End_Date__c = camp.Campaign_End_Date__c;
+				sch.Evaluation_Start_Date__c = sch.Start_Date__c;
+				sch.Evaluation_End_Date__c = sch.End_Date__c.addDays(ConstantUtil.EVALUATION_DATE_OFFSET);
+				schs.add(sch);
+			}
+		}
+		insert schs;
+	}
+	
+	public void cascadeDeleteSchedulesOnUpdate(List<CW_Campaign__c> newObjects, Map<Id,CW_Campaign__c> oldObjectMap)
+	{
+		List<CW_Campaign__c> camps = new List<CW_Campaign__c>();	// list of campaigns which need schedules removed
+		for(CW_Campaign__c camp : newObjects)
+		{
+			if(camp.Recurrence_Type__c != oldObjectMap.get(camp.id).Recurrence_Type__c ||
+			   camp.Campaign_Start_Date__c != oldObjectMap.get(camp.id).Campaign_Start_Date__c ||
+			   camp.Campaign_End_Date__c != oldObjectMap.get(camp.id).Campaign_End_Date__c)
+			{
+				camps.add(camp);
+			}
+		}
+		
+		if(camps.isEmpty())
+		{
+			return;
+		}
+				
+		// remove Schedule + AGS + AGM, Schedule trigger will remove Messaging and AGS Reward + AGS Member
+		delete [select id from CW_Schedule__c where Marketing_Campaign__c in :camps];
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.